### DISCUSSION TITLE
Push sensor readings to prometheus

### DISCUSSION THREAD

**Discussion by iwishiwasaneagle** [arendst/Tasmota#14367] [Q&A] [answered]
As the title says. Prometheus now has the ability to receive data, rather than just scrape https://prometheus.io/docs/prometheus/latest/feature_flags/#remote-write-receiver. 

Would this be possible to implement? I'm new to Tasmota but would be willing to try and implement it if it is possible.

**Comment by sfromis** [comment]
Tasmota has since 2020 (version 8.4) had the ability to [send data](https://github.com/arendst/Tasmota/blob/development/tasmota/xsns_75_prometheus.ino#L21-L33) to Prometheus, if you build with that optional feature. Don't ask me about details ;-)

**Reply by iwishiwasaneagle** [reply]
This isn't pushing any data, it just allows prometheus scrape from `[IP]/metrics` at a defined interval.

```bash
$ curl http://192.168.0.93/metrics
# TYPE tasmota_info gauge
tasmota_info{version="10.1.0",image="(tasmota)",build_timestamp="2022-01-10T16:56:40",devicename="Tasmota"} 1
# TYPE tasmota_uptime_seconds gauge
tasmota_uptime_seconds{} 97
# TYPE tasmota_boot_count counter
tasmota_boot_count{} 21
# TYPE tasmota_flash_writes_total counter
tasmota_flash_writes_total{} 40
# TYPE tasmota_memory_size_bytes gauge
tasmota_memory_size_bytes{memory="heap"} 288660
# TYPE tasmota_memory_free_bytes gauge
tasmota_memory_free_bytes{memory="heap"} 161108
# TYPE tasmota_memory_max_alloc_bytes gauge
tasmota_memory_max_alloc_bytes{memory="heap"} 65424
# TYPE tasmota_energy_voltage_volts gauge
tasmota_energy_voltage_volts{} 0
# TYPE tasmota_energy_current_amperes gauge
tasmota_energy_current_amperes{} 0.000
# TYPE tasmota_energy_power_active_watts gauge
tasmota_energy_power_active_watts{} 0
# TYPE tasmota_energy_power_kilowatts_daily counter
tasmota_energy_power_kilowatts_daily{} 0.000
# TYPE tasmota_energy_power_kilowatts_total counter
tasmota_energy_power_kilowatts_total{} 0.000
# TYPE tasmota_sensors_a1_ gauge
tasmota_sensors_a1_{sensor="analog"} 1808
# TYPE tasmota_sensors_a2_ gauge
tasmota_sensors_a2_{sensor="analog"} 499
# TYPE tasmota_sensors_temperature_celsius gauge
tasmota_sensors_temperature_celsius{sensor="esp32"} 56.7
```

**Reply by barbudor** [reply]
Similar feature has been implemented for InfluxDB, so why not for Prometheus too.
However there is no Master Plan which include this feature so feel free to implement it and share through a PR so it would be welcomed to be included in the base code.

**Comment by MarkusTeufelberger** [comment, marked as answer]
Prometheus _really_ isn't built for getting data pushed towards it. If you need this feature, push to an InfluxDB (directly or through MQTT) and use this as a remote read target in Prometheus instead (https://docs.influxdata.com/influxdb/v1.8/supported_protocols/prometheus/, https://prometheus.io/docs/prometheus/latest/configuration/configuration/#remote_read). The docs you linked is just for data in Prometheus' own format, I doubt that it would be easy/feasible to implement all of this on a microcontroller (protobuf, snappy, gRPC, HTTP2... probably some crypto on top as well). It also is against the system design of Prometheus which is strongly biased towards pulling rather than pushing (https://prometheus.io/docs/introduction/faq/#why-do-you-pull-rather-than-push).
